### ISSUE TITLE
Chrome: Scripts load inconsistently on very-fresh tabs(?)

### ISSUE THREAD

**Issue by marcustyphoon** [AprilSylph/XKit-Rewritten#636]
### Platform

MacOS 10.14.6 x86

### Browser

Edge 102.0.1245.41

### Addon version

f318547

### Details

I am mostly at a loss here.

On a completely inconsistent basis, I sometimes open Tumblr to find that notificationblock is not active until a hard refresh. I finally got this to occur after adding logging before and after the dynamic import statements in main.js and at the top of every utility function, and discovered that it's actually any script that imports meatballs.js, as it for some reason is not resolved.

<img width="1265" alt="Screen Shot 2022-06-16 at 10 08 35 AM" src="https://user-images.githubusercontent.com/8336245/174127526-08b8096d-7319-4fc7-8d7c-13998ee83fcc.png">

<img width="1536" alt="Screen Shot 2022-06-16 at 10 15 36 AM" src="https://user-images.githubusercontent.com/8336245/174128965-f3a77707-dc59-4fc7-be3b-a693e72b8882.png">

I have never had this problem with a different set of scripts, but I can't figure out what would be different about the meatballs utility based on its dependencies. I also don't know if this occurs in production or if it's just a weird artifact of my local install, but I have no reproduction method yet so I can't test that really.

**Comment by marcustyphoon** [comment]
Wait... mutual checker didn't load in this screenshot either.

Uh, what?

**Comment by AprilSylph** [comment]
The current commit isn't released, so it by definition would not occur in production right now. I don't know enough about your local install, but if you hard-reset to master and reload the extension in the browser, you should get the exact same experience as if it were released.

If you can figure out how to reliably reproduce it (even if it is just "hard-refresh Tumblr x times until it happens"; I narrowed down a `fakeImport` bug that way), please rewind to 98af3c5e1706d82b647d6885a1aaef7fdf78cb6a and see if it still occurs.

I'm pausing all other development until this is resolved. If 225f7fec28fc50dd2155101269a3dc813c7b354c needs to be reverted I want a clean revert.

Also, to help me test this, do you mind sharing the config you're using?

**Comment by AprilSylph** [comment]
Okay, if 225f7fec28fc50dd2155101269a3dc813c7b354c can be ruled out as the cause, that's good. Ignore my comment about other development being paused.

**Comment by AprilSylph** [comment]
Hmm, two tabs opened at once... I know Chromium handles unfocused tabs differently. What happens if you open a Tumblr tab and let it load entirely out of focus?

On Edge (Windows 10) you can do so by CTRL+clicking a link; on Mac I assume this would be CMD+click.

**Comment by marcustyphoon** [comment]
I tried that, didn't seem to work. The one time I was hopeful I'd gotten it to happen I think I command-shift-t'd to reopen a window that had two Tumblr tabs and did click both of them, but that hadn't worked since.

I did, though, just have it happen again on a simple refresh. Mutual checker did load this time, but the meatballs utility (and scripts relying on it) did not. Trying a Firefox install and a Chrome install with no other extensions now.

**Comment by marcustyphoon** [comment]
Aha! Got a repro finally. It happens when I a) navigate to tumblr really quickly after opening a new tab or window, and b) only when I have XKit 7 enabled (??). I have `Enabled extensions: xkit_main, xkit_patches, xkit_preferences, xkit_updates, quick_tags, editable_reblogs` and `Disabled extensions: xwidgets, show_more, classic_tags, old_stats, estufars_sidebar_fix, old_blue, notifications_plus, messaging_tweaks, mass_plus, dont_stretch_photosets, disable_search, themes_plus, servant, tagviewer, mute, shorten_posts, profiler, one_click_postage, notificationblock, postblock, outbox, bookmarker, timestamps, hideavatars, highlighter, tweaks, classic_header, accesskit, fresh_prince, pokes, blacklist, anti_capitalism, audio_downloader, audio_plus, autoscroll, auto_tagger, disable_gifs, glowing_follow, go_to_dash, read_posts, panaroma, open_in_new_tabs, norecommended, titles, tf2_reblogs, vanilla_video, xcloud, xneko, retags, tag_replacer`. Haven't gotten it to happen in my Chrome/Firefox installs yet, but I haven't imported my XK7 data into them yet to see if it's specific to that.

Also, apparently meatballs.js isn't necessarily the culprit. (Why it would always be NotificationBlock, then, I have absolutely no idea about yet.)

<img width="1265" alt="Screen Shot 2022-06-16 at 7 06 39 PM" src="https://user-images.githubusercontent.com/8336245/174208925-a1ab4ee6-d29e-4d59-8db0-92961a79664b.png">

<img width="1265" alt="Screen Shot 2022-06-16 at 7 10 08 PM" src="https://user-images.githubusercontent.com/8336245/174209946-53783d72-a011-40a8-9c68-0206391d3d52.png">

**Comment by marcustyphoon** [comment]
Reproduced in Chrome. This is not my browser being cursed, phew.

**Comment by marcustyphoon** [comment]
It does appear to not fail if main.js is modified to import all of the scripts sequentially instead of simultaneously, but that means waiting on things like the user info API call completing before continuing to parse js files, and it doesn't really address whatever is actually going on here.

**Comment by marcustyphoon** [comment]
I have disabled XKit 7 (so I guess it's not that) and am now encountering this with Show Originals.

**Comment by marcustyphoon** [comment]
Failed to reproduce on a Windows PC and another Mac. Maybe my system is just cursed.

**Comment by marcustyphoon** [comment]
Some sort of interesting tests:
```diff
  const importWithRetry = async function (path) {
    const result = await Promise.race([
      import(path),
      new Promise(resolve => setTimeout(() => resolve(undefined), 5000))
    ]);
    console.log(result ? `loaded ${path}` : `retrying ${path}`);
    return result ?? importWithRetry(path);
  };

-   const { main, clean, stylesheet, onStorageChanged } = await import(scriptPath);
+   const { main, clean, stylesheet, onStorageChanged } = await importWithRetry(scriptPath); 
```
This fix doesn't work. For the scripts that fail (which ones depend on which are enabled), it prints "retrying" forever; apparently once this bug is triggered on a file import that file will always fail to import.

```diff
-   installedScripts
-     .filter(scriptName => enabledScripts.includes(scriptName))
-     .forEach(runScript);
- 
+   for (const scriptName of installedScripts.filter(scriptName => enabledScripts.includes(scriptName))) {
+     await runScript(scriptName);
+   }
```
This fix, not importing all of the scripts at once, works. (With a penalty to extension load time, I assume).

```diff
-   installedScripts
-     .filter(scriptName => enabledScripts.includes(scriptName))
-     .forEach(runScript);
- 
+   const sleep = ms => new Promise(resolve => setTimeout(() => resolve(), ms));
+   for (const scriptName of installedScripts.filter(scriptName => enabledScripts.includes(scriptName))) {
+     await sleep(0);
+     runScript(scriptName);
+   }
```
This fix, importing all of the scripts in different function calls but allowing them to be simultaneous, mostly works but if I recall correctly it caused some tweaks to occasionally suffer the same problem unless the fix was also applied to Tweaks (and presumably No Recommended, etc).

```diff
+   const sleep = ms => new Promise(resolve => setTimeout(() => resolve(), ms));
+   await sleep(5000);

    installedScripts
      .filter(scriptName => enabledScripts.includes(scriptName))
```
This fix, simply delaying the unmodified simultaneous imports, works if the timeout is high enough, but not if it is very low; apparently, either the amount of other code running at the time or some specific piece of Tumblr code that runs at the time is important. It is not just the act of simultaneously importing all of these scripts.

**Comment by AprilSylph** [comment]
> This fix, not importing all of the scripts at once, works. (With a penalty to extension load time, I assume).

I think we should measure this. Also, calling `runScript` sequentially is not the only option here - we could do `await import` on the file and then call `runScript` as normal, assuming it _is_ just the import tree that's the problem.

**Comment by marcustyphoon** [comment]
> I think we should measure this.

I haven't measured it numerically but it's definitely nonzero, as it reorders all but the first few extension loads to come after a bunch of Redpop rendering stuff.

> we could do await import on the file and then call runScript as normal,

Yep. This has roughly the same effect on when everything actually gets run, I think.

Another option that seemed to work in a brief test was changing `run_at` back to `document_idle`, but I haven't looked at what that does to speed.

Anyway, I still haven't gotten this to occur on another computer, so while it obviously concerns me that it might be happening to a user and just making them think the extension is haunted, I don't feel like I've definitely confirmed that it's an issue yet.

**Comment by marcustyphoon** [comment]
Reproduced on another PC (Edge 102.0.1468.62, Windows 7, 2.4GHz Core 2 Duo, 4GB RAM). Yes, I just grabbed the oldest laptop I could find.

Sequentially calling `await import` before "simultaneously" running `runScript` does work.

**Comment by marcustyphoon** [comment]
This is fixed (without functional drawback) by #897, though it's a lot of fairly unintuitive code just to do this.
